File: src/abc299/d.py


# S = "0010011010111001"

###
# def oracle(qu):
#     return int(S[qu-1])
    

###
n = int(input())
# n = len(S)

n_loop = 20

i_zero = 1
i_one = n

for _ in range(n_loop):
    qu = (i_zero + i_one) // 2
    print("?", qu)
    s = int(input())
    # s = oracle(qu)

    if s==0:
        i_zero = qu
    elif s==1:
        i_one = qu
    

    if i_one-i_zero == 1:
        break

print("!", i_zero)

# print(S[i_zero-1], S[i_zero])

File: src/abc299/d.rs
use proconio::input;
use std::io::{stdout, Write};

// static S: &str = "000111100110011";
// fn oracle(qu: usize)->u8 {
//     let ss: Vec<char> = S.chars().collect();
//     let s: u8 = ss[qu as usize -1].to_digit(10).unwrap() as u8;
//     return s
// }

fn main() {
    input! {
        n: usize,
    }
    // let n: usize = S.len();
    let n_loop: usize = 20;

    let mut i_zero: usize = 1;
    let mut i_one: usize = n;

    for _i in 0..n_loop {
        let qu: usize = (i_zero + i_one) / 2;
        println!("? {}", qu);
        stdout().flush().unwrap();

        input! {
            s: u8,
        }
        // let s: u8 = oracle(qu);
        // println!("ans {}", s);

        match s {
            0 => {
                i_zero = qu;
            },
            1 => {
                i_one = qu;
            },
            _ => (),
        }

        if i_one-i_zero == 1 {
            break
        }
    }
    println!("! {}", i_zero);
    stdout().flush().unwrap();
    
    // let res0 = oracle(i_zero);
    // let res1 = oracle(i_one);
    // println!("{} {}", res0, res1);
}
